File: examples/contrib/magic_square.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using Google.OrTools.ConstraintSolver;

public class MagicSquare
{
    /**
     *
     * Solves the Magic Square problem.
     * See http://www.hakank.org/or-tools/magic_square.py
     *
     */
    private static void Solve(int n = 4, int num = 0, int print = 1)
    {
        Solver solver = new Solver("MagicSquare");

        Console.WriteLine("n: {0}", n);

        //
        // Decision variables
        //
        IntVar[,] x = solver.MakeIntVarMatrix(n, n, 1, n * n, "x");
        // for the branching
        IntVar[] x_flat = x.Flatten();

        //
        // Constraints
        //
        long s = (n * (n * n + 1)) / 2;
        Console.WriteLine("s: " + s);

        IntVar[] diag1 = new IntVar[n];
        IntVar[] diag2 = new IntVar[n];
        for (int i = 0; i < n; i++)
        {
            IntVar[] row = new IntVar[n];
            for (int j = 0; j < n; j++)
            {
                row[j] = x[i, j];
            }
            // sum row to s
            solver.Add(row.Sum() == s);

            diag1[i] = x[i, i];
            diag2[i] = x[i, n - i - 1];
        }

        // sum diagonals to s
        solver.Add(diag1.Sum() == s);
        solver.Add(diag2.Sum() == s);

        // sum columns to s
        for (int j = 0; j < n; j++)
        {
            IntVar[] col = new IntVar[n];
            for (int i = 0; i < n; i++)
            {
                col[i] = x[i, j];
            }
            solver.Add(col.Sum() == s);
        }

        // all are different
        solver.Add(x_flat.AllDifferent());

        // symmetry breaking: upper left is 1
        // solver.Add(x[0,0] == 1);

        //
        // Search
        //

        DecisionBuilder db = solver.MakePhase(x_flat, Solver.CHOOSE_FIRST_UNBOUND, Solver.ASSIGN_CENTER_VALUE);

        solver.NewSearch(db);

        int c = 0;
        while (solver.NextSolution())
        {
            if (print != 0)
            {
                for (int i = 0; i < n; i++)
                {
                    for (int j = 0; j < n; j++)
                    {
                        Console.Write(x[i, j].Value() + " ");
                    }
                    Console.WriteLine();
                }
                Console.WriteLine();
            }

            c++;
            if (num > 0 && c >= num)
            {
                break;
            }
        }

        Console.WriteLine("\nSolutions: {0}", solver.Solutions());
        Console.WriteLine("WallTime: {0}ms", solver.WallTime());
        Console.WriteLine("Failures: {0}", solver.Failures());
        Console.WriteLine("Branches: {0} ", solver.Branches());

        solver.EndSearch();
    }

    public static void Main(String[] args)
    {
        int n = 4;
        int num = 0;
        int print = 1;

        if (args.Length > 0)
        {
            n = Convert.ToInt32(args[0]);
        }

        if (args.Length > 1)
        {
            num = Convert.ToInt32(args[1]);
        }

        if (args.Length > 2)
        {
            print = Convert.ToInt32(args[2]);
        }

        Solve(n, num, print);
    }
}

File: examples/contrib/magic_square.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""

  Magic squares in Google CP Solver.

  Magic square problem.

  This model was created by Hakan Kjellerstrand (hakank@gmail.com)
  Also see my other Google CP Solver models:
  http://www.hakank.org/google_or_tools/
"""
import sys
from ortools.constraint_solver.python import constraint_solver as cp


def main(n, limit):
  # Create the solver.
  solver = cp.Solver("n-queens")
  if not solver:
    return

  #
  # data
  #

  #
  # declare variables
  #
  x = {}
  for i in range(n):
    for j in range(n):
      x[(i, j)] = solver.new_int_var(1, n * n, "x(%i,%i)" % (i, j))
  x_flat = [x[(i, j)] for i in range(n) for j in range(n)]

  # the sum
  # s = ( n * (n*n + 1)) / 2
  s = solver.new_int_var(1, n * n * n, "s")

  #
  # constraints
  #
  # solver.add(s == ( n * (n*n + 1)) / 2)

  solver.add_all_different(x_flat)

  [solver.add(solver.sum([x[(i, j)] for j in range(n)]) == s) for i in range(n)]
  [solver.add(solver.sum([x[(i, j)] for i in range(n)]) == s) for j in range(n)]

  solver.add(solver.sum([x[(i, i)] for i in range(n)]) == s)  # diag 1
  solver.add(solver.sum([x[(i, n - i - 1)] for i in range(n)]) == s)  # diag 2

  # symmetry breaking
  # solver.add(x[(0,0)] == 1)

  #
  # solution and search
  #
  solution = solver.assignment()
  solution.add(x_flat)
  solution.add(s)

  # db: DecisionBuilder
  db = solver.phase(
      x_flat,
      # cp.IntVarStrategy.INT_VAR_DEFAULT,
      cp.IntVarStrategy.CHOOSE_FIRST_UNBOUND,
      # cp.IntVarStrategy.CHOOSE_MIN_SIZE_LOWEST_MAX,

      # cp.IntValueStrategy.ASSIGN_MIN_VALUE
      cp.IntValueStrategy.ASSIGN_CENTER_VALUE)

  solver.new_search(db)
  num_solutions = 0
  while solver.next_solution():
    print("s:", s.value())
    for i in range(n):
      for j in range(n):
        print("%2i" % x[(i, j)].value(), end=" ")
      print()

    print()
    num_solutions += 1
    if num_solutions > limit:
      break
  solver.end_search()

  print()
  print("num_solutions:", num_solutions)
  print("failures:", solver.num_failures)
  print("branches:", solver.num_branches)
  print("WallTime:", solver.wall_time_ms)


n = 4
limit=100
if __name__ == "__main__":
  if len(sys.argv) > 1:
    n = int(sys.argv[1])
  if len(sys.argv) > 2:
    limit = int(sys.argv[2])

  main(n, limit)
